[export] allow static constraints in dynamic_shapes

pianpwk opened the pull request:
This PR allows users to specify int values for dimensions in dynamic_shapes as well as None, for example:

```
class Foo(torch.nn.Module):
    def forward(self, x, y, z):
        ...

    foo = Foo()
    inputs = (torch.randn(4, 6), torch.randn(5, 4), torch.randn(3, 3))

for dynamic_shapes in [
    None
    ((4, 6), (5, 4), (3, 3)),
    ((None, 6), None, {0: 3, 1: 3})
]:
    _ = export(foo, inputs, dynamic_shapes=dynamic_shapes)
```

All of the above should produce the same ExportedProgram.

This is done by temporarily creating a static dim constraint during analysis, where vr.lower == vr.upper. These constraints are then deleted during _process_constraints(), and do not show up in the final ExportedProgram's range_constraints.

Additionally, export() will also fail if the shapes are mis-specified, for example:
```
_ = export(foo, inputs, dynamic_shapes=((5, None), None, None))
```
leads to `torch._dynamo.exc.UserError: Static shape constraint of 5 does not match input size of 4, for L['x'].size()[0]`

avikchaudhuri commented on torch/export/dynamic_shapes.py:
```diff
@@ -698,9 +698,16 @@ def check_same_bounds(dim):
     def update_symbols(tensor, shape):
+        def _create_static_dim(tensor, i, value):
+            return _Dim(
+                f"static_dim_{id(tensor)}_{i}", (int,), {"min": value, "max": value}
```
Hmm so this is not what I had in mind. I think it would be better to have a subclass of `_Dim` to represent this case. You can probably define min and max as properties that give you the same constant. Or you could explicitly handle this subclass wherever it appears (e.g., look for `_DerivedDim`). 

In general I don't want to violate the assumption that `min < max`, I know this is an easy way to implement this but it might cause problems with `ValueRanges` assumptions elsewhere in `symbolic_shapes`. Open to be convinced otherwise.

avikchaudhuri commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -2630,7 +2630,8 @@ def create_symbol(
                 raise ConstraintViolationError(
-                    f"Static shape constraint of {constraint_dim.vr.lower} does not match input size of {val}, for {source.name()}"
+                    f"Static shape constraint of {constraint_dim.vr.lower} does not match input size of {val}, "
+                    f"for input {source.name()}, dim {source.idx}"
```
Thanks! Example of what `source.name()` gives? Why do you need `source.idx`?

pianpwk committed:
[export] allow static constraints in dynamic_shapes (#121860)

Summary:
This PR allows users to specify int values for dimensions in dynamic_shapes as well as None, for example:

```
class Foo(torch.nn.Module):
    def forward(self, x, y, z):
        ...

    foo = Foo()
    inputs = (torch.randn(4, 6), torch.randn(5, 4), torch.randn(3, 3))

for dynamic_shapes in [
    None
    ((4, 6), (5, 4), (3, 3)),
    ((None, 6), None, {0: 3, 1: 3})
]:
    _ = export(foo, inputs, dynamic_shapes=dynamic_shapes)
```

All of the above should produce the same ExportedProgram.

This is done by temporarily creating a static dim constraint during analysis, where vr.lower == vr.upper. These constraints are then deleted during _process_constraints(), and do not show up in the final ExportedProgram's range_constraints.

Additionally, export() will also fail if the shapes are mis-specified, for example:
```
_ = export(foo, inputs, dynamic_shapes=((5, None), None, None))
```
leads to `torch._dynamo.exc.UserError: Static shape constraint of 5 does not match input size of 4, for L['x'].size()[0]`


Reviewed By: avikchaudhuri

Differential Revision: D54924742

Pulled By: pianpwk
